# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/basic/BasicCanvasLayer.cpp
# Commit: dd00e2da3a0f
# Full Hash: dd00e2da3a0f11a7cc0af9e5ece21a339f4393eb
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-06-11 04:04:33
# Regressor Bug: 1632249
# File Overlap Count: 3
# Description:
#   Bug 1632249 - Support compositing out-of-process WebGL. r=handyman,lsalzman,nical,geckoview-reviewers,agi,imanol
#   
#   * Majorly simplity CanvasRenderer
#   * Replace GLScreenBuffer with trivial GLSwapChain
#   * Use descriptor structs so that future SharedSurface changes aren't so painful
# ==============================================================================

diff -r 69cc1de6f754 -r dd00e2da3a0f gfx/layers/basic/BasicCanvasLayer.cpp
--- a/gfx/layers/basic/BasicCanvasLayer.cpp	Wed Jun 10 21:57:19 2020 +0000
+++ b/gfx/layers/basic/BasicCanvasLayer.cpp	Wed Jun 10 22:21:02 2020 +0000
@@ -5,11 +5,11 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include "BasicCanvasLayer.h"
-#include "AsyncCanvasRenderer.h"
 #include "basic/BasicLayers.h"      // for BasicLayerManager
 #include "basic/BasicLayersImpl.h"  // for GetEffectiveOperator
-#include "CopyableCanvasRenderer.h"
+#include "CanvasRenderer.h"
 #include "mozilla/mozalloc.h"  // for operator new
+#include "mozilla/Maybe.h"
 #include "nsCOMPtr.h"          // for already_AddRefed
 #include "nsISupportsImpl.h"   // for Layer::AddRef, etc
 #include "gfx2DGlue.h"
@@ -30,33 +30,19 @@
 void BasicCanvasLayer::Paint(DrawTarget* aDT, const Point& aDeviceOffset,
                              Layer* aMaskLayer) {
   if (IsHidden()) return;
-
-  RefPtr<SourceSurface> surface;
-  CopyableCanvasRenderer* canvasRenderer =
-      mCanvasRenderer->AsCopyableCanvasRenderer();
-  MOZ_ASSERT(canvasRenderer);
-  if (IsDirty()) {
-    Painted();
-
-    surface = canvasRenderer->ReadbackSurface();
-  }
+  // Ignore IsDirty().
 
-  bool bufferPoviderSnapshot = false;
-  PersistentBufferProvider* bufferProvider =
-      canvasRenderer->GetBufferProvider();
-  if (!surface && bufferProvider) {
-    surface = bufferProvider->BorrowSnapshot();
-    bufferPoviderSnapshot = !!surface;
-  }
+  const auto& cr = mCanvasRenderer;
+  MOZ_ASSERT(cr);
+  const auto snapshot = cr->BorrowSnapshot();
+  if (!snapshot) return;
+  const auto& surface = snapshot->mSurf;
 
-  if (!surface) {
-    return;
-  }
-
-  Matrix oldTM;
-  if (canvasRenderer->NeedsYFlip()) {
-    oldTM = aDT->GetTransform();
-    aDT->SetTransform(Matrix(oldTM)
+  Maybe<Matrix> oldTM;
+  if (!cr->YIsDown()) {
+    // y-flip
+    oldTM = Some(aDT->GetTransform());
+    aDT->SetTransform(Matrix(*oldTM)
                           .PreTranslate(0.0f, mBounds.Height())
                           .PreScale(1.0f, -1.0f));
   }
@@ -67,17 +53,15 @@
       DrawOptions(GetEffectiveOpacity(), GetEffectiveOperator(this)),
       aMaskLayer);
 
-  if (canvasRenderer->NeedsYFlip()) {
-    aDT->SetTransform(oldTM);
+  if (oldTM) {
+    aDT->SetTransform(*oldTM);
   }
 
-  if (bufferPoviderSnapshot) {
-    bufferProvider->ReturnSnapshot(surface.forget());
-  }
+  Painted();
 }
 
-CanvasRenderer* BasicCanvasLayer::CreateCanvasRendererInternal() {
-  return new CopyableCanvasRenderer();
+RefPtr<CanvasRenderer> BasicCanvasLayer::CreateCanvasRendererInternal() {
+  return new CanvasRenderer();
 }
 
 already_AddRefed<CanvasLayer> BasicLayerManager::CreateCanvasLayer() {